src/main/kotlin/io/curity/identityserver/plugin/frejaeid/authentication/WaitRequestHandler.kt

```diff
@@ -16,25 +16,136 @@
+import se.curity.identityserver.sdk.errors.ErrorCode
+import se.curity.identityserver.sdk.http.HttpRequest
+import se.curity.identityserver.sdk.http.HttpResponse
+import se.curity.identityserver.sdk.http.HttpStatus
+import se.curity.identityserver.sdk.service.WebServiceClient
 import se.curity.identityserver.sdk.web.Request
 import se.curity.identityserver.sdk.web.Response
-import java.util.Optional
+import se.curity.identityserver.sdk.web.ResponseModel
+import java.net.URI
+import java.util.*
+import kotlin.collections.HashMap
 
-class WaitRequestModel(request : Request)
 
-class WaitRequestHandler : AuthenticatorRequestHandler<WaitRequestModel>
-{
-    override fun get(requestModel: WaitRequestModel, response: Response): Optional<AuthenticationResult>
-    {
-        TODO("poll Freja using the auth request transaction ID from the session")
+class WaitRequestHandler(private val config: FrejaEidAuthenticatorPluginConfig) : AuthenticatorRequestHandler<Request> {
+
+    private val _logger: Logger = LoggerFactory.getLogger(StartRequestHandler::class.java)
+
+    override fun get(requestModel: Request, response: Response): Optional<AuthenticationResult> = Optional.empty()
+
+    override fun preProcess(request: Request, response: Response): Request {
+        if (request.isGetRequest) {
+            response.setResponseModel(ResponseModel.templateResponseModel(emptyMap(), "authenticate/wait"),
+                    Response.ResponseModelScope.NOT_FAILURE)
+        }
+
+        // on request validation failure, we should use the same template as for NOT_FAILURE
+        response.setResponseModel(ResponseModel.templateResponseModel(emptyMap(),
+                "authenticate/wait"), HttpStatus.BAD_REQUEST)
+        return request
     }
-    
-    override fun preProcess(request: Request, response: Response) = WaitRequestModel(request)
-    
-    override fun post(requestModel: WaitRequestModel, response: Response): Optional<AuthenticationResult>
-    {
-        TODO("polling is done and the form was submitted")
+
+    override fun post(requestModel: Request, response: Response): Optional<AuthenticationResult> {
+        val responseData: Map<String, String> = checkAuthenticationStatus() as Map<String, String>
+        if (responseData["status"] == "APPROVED") {
+
+            val jwtParts = Objects.toString(responseData["details"]).split("\\.".toRegex(), 3).toTypedArray()
+
+            if (jwtParts.size < 2) {
+                throw config.exceptionFactory.internalServerException(ErrorCode.EXTERNAL_SERVICE_ERROR, "Invalid JWS")
+            }
+
+            val base64Url = Base64.getUrlDecoder()
+            val body = String(base64Url.decode(jwtParts[1]))
+            val claimsMap = config.json.fromJson(body)
+
+            val subjectAttributes = ArrayList<Attribute>()
+
+            if (config.userInfoType == UserInfoType.SSN) {
+                val basicUserInfo: HashMap<String, String>? = when {
```

Comment:
rather than nullable map, better to make an empty one if possible. Perhaps it doesn't matter though since this may be removed shortly.

Reply:
yes